src/internal/operators/share.ts

```diff
@@ -90,25 +95,44 @@ export function share<T>(options: ShareConfig<T>): MonoTypeOperatorFunction<T>;
  *
  * @return A function that returns an Observable that mirrors the source.
  */
-export function share<T>(options?: ShareConfig<T>): OperatorFunction<T, T> {
-  options = options || {};
-  const { connector = () => new Subject<T>(), resetOnComplete = true, resetOnError = true, resetOnRefCountZero = true } = options;
+export function share<T>(options: ShareConfig<T> = {}): MonoTypeOperatorFunction<T> {
+  const { connector = () => new Subject<T>() } = options;
 
   let connection: Subscription | null = null;
+  let resetConnection: Subscription | null = null;
   let subject: SubjectLike<T> | null = null;
   let refCount = 0;
   let hasCompleted = false;
   let hasErrored = false;
 
+  const cancelReset = () => {
+    resetConnection?.unsubscribe();
+    resetConnection = null;
+  };
   // Used to reset the internal state to a "cold"
   // state, as though it had never been subscribed to.
   const reset = () => {
+    cancelReset();
     connection = subject = null;
     hasCompleted = hasErrored = false;
   };
+  const resetAndUnsubscribe = () => {
+    // We need to capture the connection before
+    // we reset (if we need to reset).
+    const conn = connection;
+    reset();
+    conn?.unsubscribe();
+  };
+
+  const resetOnComplete = createResetNotifierFactory(reset, options.resetOnComplete);
+  const resetOnError = createResetNotifierFactory(reset, options.resetOnError);
+  const resetOnRefCountZero = createResetNotifierFactory(resetAndUnsubscribe, options.resetOnRefCountZero);
```

Comment:
Question: do we want to pay the normalization cost once up front or should I change this and normalize when the reset happens?

Reply:
i have moved the code to lazy normalization